hypothesis-python/src/hypothesis/searchstrategy/types.py

```diff
@@ -117,10 +119,17 @@ def from_typing_type(thing):
         return st.tuples(*map(st.from_type, elem_types))
     if isinstance(thing, typing.TypeVar):
         if getattr(thing, "__bound__", None) is not None:
-            return st.from_type(thing.__bound__)
+            strat = unwrap_strategies(st.from_type(thing.__bound__))
+            if not isinstance(strat, OneOfStrategy):
+                return strat
+            # The bound was a union, or we resolved it as a union of subtypes,
+            # so we need to unpack the strategy to ensure consistency across uses.
+            return st.shared(
+                st.sampled_from(strat.original_strategies), key="typevar=%r" % (thing,)
+            ).flatmap(lambda s: s)
```

Comment:
This looks good, but I can't quite figure out what this `shared().flatmap(lambda x: x)` is all about

Reply:
`shared()` ensures that we make the same choice every time the strategy is drawn from in a single test case. Since we're drawing a strategy but want a value, the flatmap adds another draw outside the shared - net result, distinct values of a single type!

I'll expand the comment to that effect.